daffodil-core/src/main/scala/org/apache/daffodil/dpath/Expression.scala

```diff
@@ -943,155 +950,148 @@ sealed abstract class StepExpression(val step: String, val pred: Option[Predicat
     NodeInfo.ArrayIndex
   }
 
+}
+
+sealed abstract class DownStepExpression(s: String, predArg: Option[PredicateExpression])
+  extends StepExpression(s, predArg) {
+
   override lazy val inherentType: NodeInfo.Kind = {
     if (!isLastStep) NodeInfo.Complex
     else {
-      if (stepElement.optPrimType.isDefined) {
+      if (stepElements.head.optPrimType.isDefined) {
```

Comment:
What if the elements have a different primType. At the moment, I believe we allow for inherentType to be pretty much anything, and just have the compiler add a cast if needed. Eg. if we have { ../foo + 1 }, where foo is a string type, I believe we would cast ../foo to some numeric type. We could have a scenerio where ../foo is a string in one context, but an int in another context. For a more reasonable example, we might have ../foo have primType decimal in one context and primType double in another (and, perhaps, primType float in yet another).

Reply:
Right, so this is an issue where the expression such as ../foo is polymorphic.  We either have to support that, or rule it out. The easiest and conservative design choice is to rule it out, so I will add a check that all uses are the same primType.

Reply:
Ah. There are lots of polymorphic possibilities. 

fn:count(pathToArray) cares only that the path is to an array, but does not care if in some cases the path is an array of simple type, another case path is an array of complex type. 

There are functions where only the node matters, not its type. E.g., fn:local-name(path) or fn:namespace-uri(path) which work for any node at all. 

Or fn:nilled(path) which works for either simple or complex type elements since both can be nillable.

Also functions dfdl:contentLength(path, units) and dfdl:valueLength(path, units) are both potentially polymorphic in that they measure size, but don't care what the thing actually is. 

dfdl:checkConstraints(path) works for any simple type element. 

fn:exists(node) only cares about the node existing, doesn't care about the type.

So really to check consistency we have to do this in a function-specific manner, or we can insist on uniformity even when the functions being used don't strictly speaking require it.

Reply:
Are we still able to make the conservative argument? Daffodil already has an established behavior, and that behavior is required by the spec (at least for implementations that choice to implement DPath). This sounds to me like we will be deliberately introducing a regression, and it is not clear to me we have a compelling reason to do so. Further, the broken behavior is just weird from a user standpoint. You can define a valid type T, and use it at location A xor location B. That type of restriction does not strike me as reasonable.

I see a few ways around this problem without introducing a regression:

A) Detect when such a situation would arise and specialize the expression/term/type/etc as many times as needed. Eg. If we have an element <bar> that references { ../foo } at 3 locations where foo is a string, and 5 where for is an int, then we compile 2 instances of <bar>. I thought this was the idea behind the memoizedApply function I saw earlier. Further, I suspect we will find this mechanism would be useful as a way of avoiding other tricky cases arising from de-duplication. Having said that, as I noted in my comment on memoizedApply, we would need to be careful to avoid overusing this mechanism.

B) Detect when such a situation would occur and add a runtime cast that can dispatch on the type (or on a flag set in an ERD object). This wouldn't add any overhead in the common case since we only need to include such a node when the types do not match.

Reply:
Good points. 

I am not sure we're really looking at a regression here. I am not sure any expressions in any DFDL schemas require any polymorphism even though they theoretically could. 

That said, I suppose the right algorithm is just to generalize the types. So if ../foo is int in some places and float in others and string in others, then the least upper bound is AnySimpleType for those. Then the DPath compiler would error if that expression appeared in a context where AnySimpleType isn't sufficient. So fn:localName(../foo) would work, but ../foo + 7 would fail to compile.

Reply:
It is still a regression in that we would be introducing a bug where one did not previously exist. The only question is if anyone will notice. You are probably correct that no one would notice; but if someone did notice, it would clearly be a bug in Daffodil (or, at least non standards compliant behavior). As Daffodil continues to grow in popularity and capabilities, it seems to me to be only a matter of time until someone tries to do this for some reason. 

At a minimum, can we make this a Not Yet Implemented SDE? That seems less confusing, both to a hypothetical user that stumbles onto this, and a developer who notices that we are emitting an SDE for legal behaviour.

Reply:
I will look at the type-generalization approach first, and go to a NYI SDE if that gets too complex for this change set. 

I actually think the type generalization won't be very complex. everything needed to do a type lower bound is already there in the nodeinfo stuff, and issuing the diagnostics is no more complex than issuing the diagnostics about type incompatibility that I was initially considering.

Reply:
Type generalization implemented. Please see the new tests which actually do things that are polymorphic of int/float and int/Complex type.